### ISSUE TITLE
ps2000aRunBlock returns PICO_SEGMENT_OUT_OF_RANGE with Picoscope 2205 MSO

### ISSUE THREAD

**Issue by elementgreen** [colinoflynn/pico-python#85]
When running the ps2000_demo.py I get a PICO_SEGMENT_OUT_OF_RANGE error returned by self.lib.ps2000aRunBlock() in _lowLevelRunBlock() of ps2000a.py.  The only change I made to ps2000_demo.py is that I uncommented ps2000a.PS2000a() and commented the ps2000.PS2000() line.

After messing around with _lowLevelRunBlock() I managed to get things to work.  However, I admit I am completely baffled as to how it fixed it.  All I did was to add a print() call prior to the call to self.lib.ps2000aRunBlock(), does not matter what value is passed to print().  Thinking this may be a timing issue I even tried a time.sleep () call prior, without the print, but it did not work.  I have no idea why adding the print() function works.  At least I can proceed forward though.  If anyone has any ideas of a proper fix for this, I'd greatly appreciate it.

Details:
Scope: PicoScope 2205MSO
OS: Ubuntu Linux 17.04

**Comment by hmaarrfk** [comment]
I'm inclined to think it is a a type error.

Are you able to go through to the types (c_types) used in the variables of the functions you called?

**Comment by elementgreen** [comment]
I checked over the types previously and just did so again.  Here is the function prototype from the PicoScope 2000 Series (A API) Programmer's Guide PDF:
PICO_STATUS ps2000aRunBlock (
short handle, long noOfPreTriggerSamples,
long noOfPostTriggerSamples, unsigned long timebase,
short oversample, long *timeIndisposedMs, unsigned short segmentIndex,
ps2000aBlockReady lpReady, void * pParameter)

And here is the call in ps2000a.py:
m = self.lib.ps2000aRunBlock(
c_int16(self.handle), c_uint32(numPreTrigSamples),
c_uint32(numPostTrigSamples), c_uint32(timebase),
c_int16(oversample), byref(timeIndisposedMs), c_uint16(segmentIndex),
c_void_p(), c_void_p())

The numPreTrigSamples and numPostTrigSamples should be c_int32, but I don't see how this would cause an issue, since using c_uint32 should be fine except for negative numbers.  I changed it to c_int32 and it was still breaking when removing the print() function call.

Reading up on byref() in the ctypes documentation I found mention of the use of pointer() instead, which is supposed to also create a pointer reference to the value passed to it, though it creates a "real pointer object", whatever that means.  When I changed the byref() call for timeIndisposedMs to pointer() instead it fixed the issue (can remove the print() call and it works fine).  Not sure why this is, but it probably has something to do with ctypes on my platform.

**Comment by hmaarrfk** [comment]
First, thank you for spending time to debug this.

This is so weird. `byref` has been used all over to return parameters before. I'm so surprised that this is the issue. I looked at the documentation, and in stack exchange a little and it seems.

I think the real difference between them is that `byref` simply gets the pointer, while `pointer` increments the number of references in the python object, which is necessary for multi threaded code. I'm saying this because of what I see here: 
https://github.com/python/cpython/blob/2.7/Modules/_ctypes/callproc.c#L1817
and
https://github.com/python/cpython/blob/2.7/Modules/_ctypes/callproc.c#L1599

Let me know if you have more issues after you do your fix. I have a strange feeling that it will cause memory errors again. pointer is simply a more intensive call and might have put things on the stack that seems like it "fixed" things, the same way the `print` statements did.

In either case, I will reitterate your fix:
1. Add import `pointer` to the `cytpes` import line
2. Change `byref` to `pointer`

Just in case I don't hear back from you in a few days, I'll include your fix then.

**Comment by elementgreen** [comment]
I set a breakpoint with gdb on the C function to get to the bottom of things.  The symbols are stripped from /opt/picoscope/lib/libps2000a.so though, so gdb doesn't properly show the function parameters.  After brushing up on AMD64 calling convention (https://en.wikipedia.org/wiki/X86_calling_conventions#System_V_AMD64_ABI) I discovered some things.

The first 6 parameters are stored in registers and have the correct function parameters for the working and non-working cases.  The additional 3 parameters are stored on the stack.  The first parameter incidentally is the segmentIndex.  This is where things differ between the non-working case:
```
(gdb) x/4gx $rsp
0x7fffffffd148:	0x00007ffff6706e18	0x00007fffffff0000
0x7fffffffd158:	0x0000000000000000	0x0000000000000000
```

and working case:
```
(gdb) x/4gx $rsp
0x7fffffffd148:	0x00007ffff6706e18	0x00007fff00000000
0x7fffffffd158:	0x0000000000000000	0x0000000000000000
```
The first 64 bit number is the function return address.  The second 64 bit number is the segmentIndex padded to 64 bits.  In the non-working case, the lower 16 bits are 0, but bits 17-31 are set.  I suspect that this C function is actually using a 32 bit integer for the segmentIndex, which means that any previous uninitialized portion of the 32 bits will cause this parameter to be interpreted incorrectly.  I changed the segmentIndex to c_uint32 instead and it seems to fix this problem.  I changed the pointer() back to byref() for timeIndisposedMs and removed the print call, and things continue to work.

So in summary, I think segmentIndex should be c_uint32 instead of c_uint16 for this function.  Seems like an issue with the PicoScope API.

**Comment by hmaarrfk** [comment]
Wow, Ok, you went VERY deep into debugging this issue.

It seems that you probably found the correct issue.

Have you reported your issue? I used to use the forum
https://www.picotech.com/support/

**Comment by elementgreen** [comment]
Yes, was just doing this :-)  I posted it to the Linux programming forum.  I'll add a link here once the post is approved.

**Comment by hmaarrfk** [comment]
Sigh, ok, I went through and found the ps2000aApi.h file, it is located `C:\Program Files\Pico Technology\SDK\inc`

And this is the function prototype
```c
PREF0 PREF1 PICO_STATUS PREF2 PREF3 (ps2000aRunBlock)
	(
		int16_t handle,
		int32_t noOfPreTriggerSamples,
		int32_t noOfPostTriggerSamples,
		uint32_t timebase,
		int16_t oversample,
		int32_t * timeIndisposedMs,
		uint32_t segmentIndex,
		ps2000aBlockReady	 lpReady,
		void	* pParameter
	);
```
I'll leave the issue. The documentation shows a different prototype with uint16_t (can't find it now though)

![image](https://user-images.githubusercontent.com/90008/29543914-6bd54584-8699-11e7-82d0-d10bdc3c17a1.png)

I'll update the python 2000a file with uint32 for now and maybe when you get a followup from Picotech I can close the issue with the correct one.

Thanks,

Mark

**Comment by elementgreen** [comment]
I had forgotten about the header files.  Looks like all of the segmentIndex parameters in the ps2000aAPI.h header are uint32_t.  This is contrary to the PDF docs which indicate 16 bit for all segmentIndex parameters for the 2000a API.  Whats more, there are at least 2 versions of the 2000a Programmer's Guide:
https://www.picotech.com/download/manuals/picoscope-2000-series-a-api-programmers-guide.pdf
https://www.picotech.com/download/manuals/ps2000apg.en-6.pdf

The first one is newer, but also has this discrepancy.

I checked the rest of the uses of c_uint16 for segmentIndex in the pico-python sources.  There is at least one additional issue and that is with the call to ps2000aGetTriggerTimeOffset in ps2000a.py.  The other 16 bit segmentIndex parameters are all in ps4000.py, which appear to actually use 16 bit parameters according to the header.

**Comment by hsmistry** [comment]
The latest version of the Programmer's Guide is available from this URL:

https://www.picotech.com/download/manuals/picoscope-2000-series-a-api-programmers-guide.pdf

A change was made to the API functions as the new PicoScope 2000B and 2000B MSO models have deep memory so the segmentIndex now uses an unsigned 32-bit (uint32_t) data type. I'll follow this up with our Technical Publications team @elementgreen as I see the data type is still incorrect. Please use the ps2000aApi.h header file for reference.

**Comment by hmaarrfk** [comment]
Thank you for the confirmation @hsmistry
